engine/src/main/java/io/camunda/zeebe/engine/processing/bpmn/gateway/InclusiveGatewayProcessor.java

```diff
@@ -0,0 +1,149 @@
+
+import io.camunda.zeebe.el.Expression;
+import io.camunda.zeebe.engine.processing.bpmn.BpmnElementContext;
+import io.camunda.zeebe.engine.processing.bpmn.BpmnElementProcessor;
+import io.camunda.zeebe.engine.processing.bpmn.behavior.BpmnBehaviors;
+import io.camunda.zeebe.engine.processing.bpmn.behavior.BpmnIncidentBehavior;
+import io.camunda.zeebe.engine.processing.bpmn.behavior.BpmnStateTransitionBehavior;
+import io.camunda.zeebe.engine.processing.common.ExpressionProcessor;
+import io.camunda.zeebe.engine.processing.common.Failure;
+import io.camunda.zeebe.engine.processing.deployment.model.element.ExecutableInclusiveGateway;
+import io.camunda.zeebe.engine.processing.deployment.model.element.ExecutableSequenceFlow;
+import io.camunda.zeebe.protocol.record.value.ErrorType;
+import io.camunda.zeebe.util.Either;
+import io.camunda.zeebe.util.buffer.BufferUtil;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+public final class InclusiveGatewayProcessor
+    implements BpmnElementProcessor<ExecutableInclusiveGateway> {
+
+  private static final String NO_OUTGOING_FLOW_CHOSEN_ERROR =
+      "Expected at least one condition to evaluate to true, or to have a default flow";
+
+  private final BpmnStateTransitionBehavior stateTransitionBehavior;
+  private final BpmnIncidentBehavior incidentBehavior;
+  private final ExpressionProcessor expressionBehavior;
+
+  public InclusiveGatewayProcessor(final BpmnBehaviors behaviors) {
+    expressionBehavior = behaviors.expressionBehavior();
+    incidentBehavior = behaviors.incidentBehavior();
+    stateTransitionBehavior = behaviors.stateTransitionBehavior();
+  }
+
+  @Override
+  public Class<ExecutableInclusiveGateway> getType() {
+    return ExecutableInclusiveGateway.class;
+  }
+
+  @Override
+  public void onActivate(
+      final ExecutableInclusiveGateway element, final BpmnElementContext activating) {
+    if (!element.getOutgoing().isEmpty() && element.getOutgoingWithCondition().isEmpty()) {
+      // when all the outgoing condition is not set
+      // the joining of the incoming sequence flows into the inclusive gateway happens in the
+      // sequence flow processor. The activating event of the inclusive gateway is written when all
+      // incoming sequence flows are taken
+      final var activated = stateTransitionBehavior.transitionToActivated(activating);
+      final var completing = stateTransitionBehavior.transitionToCompleting(activated);
+      stateTransitionBehavior
+          .transitionToCompleted(element, completing)
+          .ifRightOrLeft(
+              completed ->
+                  // fork the process processing by taking all outgoing sequence flows of the
+                  // inclusive gateway
+                  stateTransitionBehavior.takeOutgoingSequenceFlows(element, completed),
+              failure -> incidentBehavior.createIncident(failure, completing));
+    } else {
```

Comment:
`I would also like to see a test case for the 2 cases described by @ThorbenLindhauer [here](https://forum.camunda.io/t/tokens-evaluation-in-inclusive-gateway-at-join-process/9354/12), which relates to [this process model](https://forum.camunda.io/t/tokens-evaluation-in-inclusive-gateway-at-join-process/9354/14).`

Hi,@korthout this case you offered ,the outgoing sequence flows have no condition, i think this code block should keep. what you think?

Reply:
Ah I think that's unintended. [The spec is unclear whether the forking logic](https://www.omg.org/spec/BPMN/2.0.2/PDF#13.4.3%20inclusive%20Gateway%20(Inclusive%20Decision%20and%20Inclusive%20Merge))  requires conditions (see the table in this section). However, it only discusses evaluation of conditions. In addition, the exclusive gateway also requires conditions. Let's align these.

> In order to determine the outgoing Sequence Flows that receive a token, all
> conditions on the outgoing Sequence Flows are evaluated. The evaluation
> does not have to respect a certain order.
> For every condition which evaluates to true, a token MUST be passed on the
> respective Sequence Flow.
> If and only if none of the conditions evaluates to true, the token is passed on the
> default Sequence Flow.
> In case all conditions evaluate to false and a default flow has not been specified,
> the Inclusive Gateway throws an exception.

Reply:
I see, i'll change it.